srml/support/test/Cargo.toml

```diff
@@ -0,0 +1,24 @@
+[package]
```

Comment:
Why do we need an extra crate for this?

Reply:
I written it in the commit message:

this allow to test for construct_runtime as well.

The reason to have put that in another crate is:
* if we put test in `tests/` dir of srml/support then decl_storage fails to get
  srml-support access because it believes it is inside srml-support
  crate and so derive access to `quote!{ crate }` but this is wrong
  (and I don't see any way to prevent that, and it only bother us so I
  don't think that matters that much)
* if we put test inside lib.rs then contruct_runtime cannot be used
  because it call some macros that are defined with macros
  (decl_outer_event and decl_outer_origin) and thus rustc complains.

Reply:
Yeah, just fix the `crate` access. Instead of generating `crate`, use the hidden include. While generating the hidden include, change the generation to check if `CARGO_PKG_NAME` is `srml-support` and import `srml-support` by `extern crate srml-support` instead of using my `proc_macro_crate`. Just look into `parity-codec`.

Reply:
Yes we do so already:
```rust
	let scrate = generate_crate_access(&hidden_crate_name, "srml-support");
```
```rust 
/// Generates the access to the `srml-support` crate.
pub fn generate_crate_access(unique_id: &str, def_crate: &str) -> TokenStream {
	if ::std::env::var("CARGO_PKG_NAME").unwrap() == def_crate {
		quote::quote!( crate )
	} else {
		let mod_name = generate_hidden_includes_mod_name(unique_id);
		quote::quote!( self::#mod_name::hidden_include )
	}.into()
}
```

`CARGO_PKG_NAME` is `srml-support` for test in tests dir.
I don't know how we can determine that we are actually in this kind of environment.
in std::env current_dir return "/home/thiolliere/Developpement/substrate/srml/support" and args are "/home/thiolliere/Developpement/substrate/target/debug/deps/instances-edf4e94ffdc37966"

Reply:
oh actually it seems we can:

when CARGO_PKG_NAME == srml-support && arg dir doesn't start with "srml_support-"

Reply:
```
pub fn generate_crate_access(unique_id: &str, def_crate: &str) -> TokenStream {
	let mod_name = generate_hidden_includes_mod_name(unique_id);
	quote::quote!( self::#mod_name::hidden_include )
}
```
+ 
```
pub fn generate_hidden_includes(unique_id: &str, def_crate: &str) -> TokenStream {
	if ::std::env::var("CARGO_PKG_NAME").unwrap() == def_crate {
		quote::quote!(
			#[doc(hidden)]
			mod #mod_name {
				pub extern crate srml-support as hidden_include;
			}
		)
	} else {
		let mod_name = generate_hidden_includes_mod_name(unique_id);

		match crate_name(def_crate) {
			Ok(name) => {
				let name = Ident::new(&name, Span::call_site());
				quote::quote!(
					#[doc(hidden)]
					mod #mod_name {
						pub extern crate #name as hidden_include;
					}
				)
			},
			Err(e) => {
				let err = Error::new(Span::call_site(), &e).to_compile_error();
				quote!( #err )
			}
		}

	}.into()
}
```

Reply:
But then inlined test inside srml/support/src can't work anymore because they can't write `extern crate srml_support as hidden_include;`

Reply:
Hmm yeah :( With `1.34` we can start using `extern crate self`. So continue using an extra crate :)

Reply:
oh cool!

Reply:
yes we still won't be able to write `extern crate srml_support as hidden_include;`
in inline test of srml-support.
and extern crate self as hidden_include in tests dir doesn't link to srml-support but to the test itself.